README.md
# Fall Detection

[![](https://travis-ci.org/computationalcore/fall-detection.svg?branch=master)](https://travis-ci.org/computationalcore/fall-detection)

This project consists on showcasing the advantages of the Intel’s OpenVINO toolkit for inference in detecting people falling in an edge application.

This app perform single person fall detection using OpenVINO's [human-pose-estimation-0001](https://docs.openvinotoolkit.org/latest/_models_intel_human_pose_estimation_0001_description_human_pose_estimation_0001.html) pre-trained model.

To detect falls, the app uses the coordinates of the head(nose, eyes and ears), neck and shoulders positions in a frame-by-frame comparison to determine
if the person is falling.

It works with a video file input or webcam stream.

[![](http://img.youtube.com/vi/C_S4oePpTZ8/0.jpg)](https://www.youtube.com/watch?v=C_S4oePpTZ8 "Fall Detection")

[Watch Video](https://www.youtube.com/watch?v=C_S4oePpTZ8)

## Prerequisites

To run the application in this tutorial, the OpenVINO™ toolkit and its dependencies must already be installed. 

Alternatively, you can create and build the docker image provided by this repository by following these [instructions](DOCKER.md).

Installation instructions may be found at: https://software.intel.com/en-us/articles/OpenVINO-Install-Linux

When needed, the following optional hardware can be used:

- USB camera - Standard USB Video Class (UVC) camera.

- Intel® Core™ CPU with integrated graphics.

- VPU - USB Intel® Movidius™ Neural Compute Stick and what is being referred to as "Myriad"

A summary of what is needed:

### Hardware
Target and development platforms meeting the requirements described in the "System Requirements" section of the OpenVINO™ toolkit documentation which may be found at: https://software.intel.com/en-us/openvino-toolkit

Note: While writing this tutorial, an Intel® i7-8550U with Intel® HD graphics 520 GPU was used as both the development and target platform.

Optional:

- Intel® Movidius™ Neural Compute Stick

- USB UVC camera

- Intel® Core™ CPU with integrated graphics

### Software
OpenVINO™ toolkit supported Linux operating system. This tutorial was run on 64-bit Ubuntu 16.04.1 LTS updated to kernel 4.15.0-43 following the OpenVINO™ toolkit installation instructions.

- The latest OpenVINO™ toolkit installed and verified. Supported versions +2018 R4.0. (Lastest version supported 2019 R1.0.1)

- Git(git) for downloading from the GitHub repository.

- BOOST library. 

To install on Ubuntu, run

``` bash
apt-get install libboost-dev libboost-log-dev
```

### Checks
By now you should have completed the Linux installation guide for the OpenVINO™ toolkit, however before continuing, please ensure:

- That after installing the OpenVINO™ toolkit you have run the supplied demo samples

- If you have and intend to use a GPU: You have installed and tested the GPU drivers

- If you have and intend to use a USB camera: You have connected and tested the USB camera

- If you have and intend to use a Myriad: You have connected and tested the USB Intel® Movidius™ Neural Compute Stick


## Build

- Clone the repository at desired location:

``` bash
git clone https://github.com/computationalcore/fall-detection
```

- The first step is to configure the build environment for the OpenCV toolkit by sourcing the "setupvars.sh" script.

``` bash
source  /opt/intel/openvino/bin/setupvars.sh
```

- For older versions than 2019 R1, OpenVINO was installed in a different dir, run this instead:

``` bash
source  /opt/intel/computer_vision_sdk/bin/setupvars.sh
```


- Change to the top git repository:

``` bash
cd fall-detection
```

- Install other project dependencies after run openvino env

``` bash
pip install -r requirements.txt
```

## Run

To check available options run

``` bash
$ python fall_detection.py -h
usage: fall_detection.py [-h] -i INPUT [-mp {FP16,FP32}] [-l CPU_EXTENSION] [-pp PLUGIN_DIR] [-d DEVICE]

Detect a person falling from a webcam or a video file

optional arguments:
  -h, --help            show this help message and exit
  -i INPUT, --input INPUT
                        Path to video file or image. 'cam' for capturing video stream from internal camera.
  -mp {FP16,FP32}, --model_precision {FP16,FP32}
                        The precision of the human pose model. Default is 32-bit integer.
  -l CPU_EXTENSION, --cpu_extension CPU_EXTENSION
                        MKLDNN (CPU)-targeted custom layers.Absolute path to a shared library with the kernels impl.
  -pp PLUGIN_DIR, --plugin_dir PLUGIN_DIR
                        Path to a plugin folder
  -d DEVICE, --device DEVICE
                        Specify the target device to infer on; CPU, GPU, FPGA or MYRIAD is acceptable. Demo will look for a
                        suitable plugin for device specified (CPU by default)
```

### Detecting falls on a video file

(in this case I use example/demo.mp4, but it can be any other)

``` bash
python fall_detection.py -i example/demo.mp4
```

### Detecting falls on webcam
``` bash
python fall_detection.py -i cam
```

## Limitations

- It works only with a single person, in the future I can add multiple people support, but currently if there is
more than one person in the scene it will confuse the detector.
- The detector only takes into consideration the relative positions of head elements, neck and shoulders. In the future it can be improved
to consider other aspects of the human pose elements 

## Authors
Vin Busquet
* [https://github.com/computationalcore](https://github.com/computationalcore)

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details

## Changelog

For details, check out [CHANGELOG.md](CHANGELOG.md).

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

## [1.0](../../releases/tag/v1.0)
- First Version


fall_detector.py
import sys
import os
import cv2
import time
import logging as log
import numpy as np

from argparse import ArgumentParser
from openvino.inference_engine import IENetwork, IEPlugin

JOINT_COLORS = [
    (0, 0, 255),
    (0, 0, 128), (255, 255, 255), (0, 255, 0), (0, 0, 255),
    (192, 192, 192), (128, 0, 255), (0, 128, 128), (255, 255, 255),
    (128, 128, 0), (128, 128, 128), (0, 255, 0), (0, 0, 255),
    (255, 255, 0), (255, 0, 255), (0, 255, 255), (0, 0, 128)
]
JOINT_COLOR = (0, 255, 255)
POSE_PAIRS = [
    [1, 0], [1, 2], [1, 5], [2, 3], [3, 4], [5, 6], [6, 7],
    [1, 8], [8, 9], [9, 10], [1, 11], [11, 12], [12, 13],
    [0, 14], [0, 15], [14, 16], [15, 17]
]


def main():
    log.basicConfig(format="[ %(levelname)s ] %(message)s", level=log.INFO, stream=sys.stdout)
    args = build_argparser().parse_args()
    model_xml = args.model
    model_bin = os.path.splitext(model_xml)[0] + ".bin"
    # Plugin initialization for specified device and load extensions library if specified
    log.info("Initializing plugin for {} device...".format(args.device))
    plugin = IEPlugin(device=args.device, plugin_dirs=args.plugin_dir)
    if args.cpu_extension and 'CPU' in args.device:
        plugin.add_cpu_extension(args.cpu_extension)
    # Read IR
    log.info("Reading IR...")
    net = IENetwork(model=model_xml, weights=model_bin)    

    if plugin.device == "CPU":
        supported_layers = plugin.get_supported_layers(net)
        not_supported_layers = [l for l in net.layers.keys() if l not in supported_layers]
        if len(not_supported_layers) != 0:
            log.error("Following layers are not supported by the plugin for specified device {}:\n {}".
                      format(plugin.device, ', '.join(not_supported_layers)))
            log.error("Please try to specify cpu extensions library path in demo's command line parameters using -l "
                      "or --cpu_extension command line argument")
            sys.exit(1)
    
    input_blob = next(iter(net.inputs))
    out_blob = next(iter(net.outputs))
    log.info("Loading IR to the plugin...")
    exec_net = plugin.load(network=net, num_requests=2)
    # Read and pre-process input image
    n, c, h, w = net.inputs[input_blob].shape
    del net
    if args.input == 'cam':
        input_stream = 0
    else:
        input_stream = args.input
        assert os.path.isfile(args.input), "Specified input file doesn't exist"
    
    cap = cv2.VideoCapture(input_stream)
    
    # Grab the shape of the input 
    width = int(cap.get(3))
    height = int(cap.get(4))
    fps = cap.get(cv2.CAP_PROP_FPS)
    font_scale = round(height/360)
    font_thickness = round(3*font_scale)
    
    out_filename = os.path.splitext(input_stream)[0]
    out = cv2.VideoWriter(out_filename + '_output.mp4', 0x00000021, fps, (width, height))
    
    log.info('Height %s' % height)
    log.info('font_scale %s' % font_scale)

    cur_request_id = 0
    next_request_id = 1

    # Detection variablez
    head_avg_position_previous = [0,0]
    previous_head_avg_position = 0
    previous_head_detection_frame = 0    
    fall_detected = False
    last_fall_detected_frame = 0
    # Fall Detection threshold speed is depedent of image height
    fall_threshold = 0.04 * height
    framerate_threshold = round(fps/5.0)
    fall_detected_text_position = (20,round(0.15*height))
    
   
        
        # Increment frame number
        frame_number += 1

        key = cv2.waitKey(1)
        if key == 27:
            break

    # Release the out writer, capture, and destroy any OpenCV windows
    out.release()
    cv2.destroyAllWindows()
    del exec_net
    del plugin


if __name__ == '__main__':
    sys.exit(main() or 0)



fall_detection.py
import argparse
import logging as log
import os
import sys

import cv2
from exitstatus import ExitStatus

from openvino.inference_engine import IENetwork, IEPlugin

JOINT_COLORS = [
    (0, 0, 255),
    (0, 0, 128), (255, 255, 255), (0, 255, 0), (0, 0, 255),
    (192, 192, 192), (128, 0, 255), (0, 128, 128), (255, 255, 255),
    (128, 128, 0), (128, 128, 128), (0, 255, 0), (0, 0, 255),
    (255, 255, 0), (255, 0, 255), (0, 255, 255), (0, 0, 128)
]

"""
Pose Data Points

Nose 0, Neck 1, Right Shoulder 2, Right Elbow 3, Right Wrist 4,
Left Shoulder 5, Left Elbow 6, Left Wrist 7, Right Hip 8,
Right Knee 9, Right Ankle 10, Left Hip 11, Left Knee 12,
LAnkle 13, Right Eye 14, Left Eye 15, Right Ear 16,
Left Ear 17, Background 18
"""
POSE_POINTS_NUMBER = 18
POSE_PAIRS = [
    [1, 0], [1, 2], [1, 5], [2, 3], [3, 4], [5, 6], [6, 7],
    [1, 8], [8, 9], [9, 10], [1, 11], [11, 12], [12, 13],
    [0, 14], [0, 15], [14, 16], [15, 17]
]


def parse_args(parse_this=None) -> argparse.Namespace:
    """Parse user command line arguments."""
    parser = argparse.ArgumentParser(
        description="Detect a person falling from a webcam or a video file"
    )
    parser.add_argument(
        "-i",
        "--input",
        help="""
            Path to video file or image. 'cam' for capturing
            video stream from internal camera.
        """,
        required=True,
        type=str
    )
    parser.add_argument(
        "-mp",
        "--model_precision",
        help="""The precision of the human pose model.
            Default is 32-bit integer.
        """,
        choices=('FP16', 'FP32'),
        default='FP32'
    )
    parser.add_argument(
        "-l",
        "--cpu_extension",
        help="""MKLDNN (CPU)-targeted custom layers.Absolute path to a shared
            library with the kernels impl.
        """,
        type=str,
        default=None
    )
    parser.add_argument(
        "-pp",
        "--plugin_dir",
        help="Path to a plugin folder",
        type=str,
        default=None
    )
    parser.add_argument(
        "-d",
        "--device",
        help="""Specify the target device to infer on; CPU, GPU, FPGA or MYRIAD
            is acceptable. Demo will look for a suitable plugin for device
            specified (CPU by default)
        """,
        default="CPU",
        type=str
    )
    return parser.parse_args(parse_this)


def main():
    """Accept arguments and perform the inference on entry"""
    # Setup log config
    log.basicConfig(
        format="[ %(levelname)s ] %(message)s",
        level=log.INFO,
        stream=sys.stdout
    )

    # Parse args
    args = parse_args()
    log.info("Start Fall Detection")

    # Plugin initialization for specified device
    log.info("Initializing plugin for {} device...".format(args.device))
    plugin = IEPlugin(device=args.device, plugin_dirs=args.plugin_dir)
    if args.cpu_extension and 'CPU' in args.device:
        plugin.add_cpu_extension(args.cpu_extension)

    # Load model
    model_xml = os.path.join(
        os.getcwd(),
        "models",
        "human-pose-estimation-0001",
        args.model_precision,
        "human-pose-estimation-0001.xml"
    )
    model_bin = os.path.splitext(model_xml)[0] + ".bin"
    log.info("Reading IR...")
    net = IENetwork(model=model_xml, weights=model_bin)

    if plugin.device == "CPU":
        supported_layers = plugin.get_supported_layers(net)
        not_supported_layers = [
            l for l in net.layers.keys() if l not in supported_layers
        ]
        if len(not_supported_layers) != 0:
            log.error(
                """Following layers are not supported by the plugin
                for specified device {}:\n {}""".
                format(plugin.device, ', '.join(not_supported_layers))
            )
            log.error(
                """Please try to specify cpu extensions library
                    path in demo's command line parameters using -l
                    or --cpu_extension command line argument
                """
            )
            sys.exit(1)
    input_blob = next(iter(net.inputs))

    log.info("Loading IR to the plugin...")
    exec_net = plugin.load(network=net, num_requests=2)

    # Read and pre-process input image
    n, c, h, w = net.inputs[input_blob].shape
    del net
    if args.input == 'cam':
        input_stream = 0
    else:
        input_stream = args.input
        assert os.path.isfile(args.input), "Specified input file doesn't exist"

    cap = cv2.VideoCapture(input_stream)
    # Grab the shape of the input
    width = int(cap.get(3))
    height = int(cap.get(4))
    fps = cap.get(cv2.CAP_PROP_FPS)
    font_scale = round(height/360)
    font_thickness = round(3*font_scale)

    cur_request_id = 0
    next_request_id = 1

    # Fall Detection variables
    previous_head_avg_position = 0
    previous_head_detection_frame = 0
    last_fall_detected_frame = 0
    # Fall Detection threshold speed is depedent of the frame height
    fall_threshold = 0.04 * height
    framerate_threshold = round(fps/5.0)
    fall_detected_text_position = (20, round(0.15*height))

    ret, frame = cap.read()
    frame_number = 0

    out_file = None
    if args.input != 'cam':
        out_filename = os.path.splitext(input_stream)[0] + '_output.mp4'
        out_file = cv2.VideoWriter(
            out_filename,
            0x00000021,
            fps,
            (width, height)
        )
        log.info("Evaluating video file stream...")
    else:
        log.info("Evaluating webcam stream...")

    while cap.isOpened():
        ret, next_frame = cap.read()
        if not ret:
            break

        # Pre-process inputs
        in_frame = cv2.resize(next_frame, (w, h))
        in_frame = in_frame.transpose((2, 0, 1))
        in_frame = in_frame.reshape((n, c, h, w))

        # Inference
        exec_net.start_async(
            request_id=next_request_id,
            inputs={input_blob: in_frame}
        )
        if exec_net.requests[cur_request_id].wait(-1) == 0:
            # Parse detection results of the current request
            res = exec_net.requests[cur_request_id].outputs
            kp_heatmaps = res['Mconv7_stage2_L2']

            threshold = 0.5
            points = []
            head_elements_y_pos = []

            for i in range(POSE_POINTS_NUMBER):
                # confidence map of corresponding body's part.
                probMap = kp_heatmaps[0, i, :, :]

                # Find global maxima of the probMap.
                minVal, prob, minLoc, point = cv2.minMaxLoc(probMap)

                # Scale the point to fit on the original image
                x = frame.shape[1] / probMap.shape[1] * point[0]
                y = frame.shape[0] / probMap.shape[0] * point[1]

                # Add point if the probability is greater than the threshold
                if prob > threshold:
                    point = (int(x), int(y))

                    # If point is a component of the head (including neck and
                    # sholders) append to the header elemnts
                    if (
                        i == 0 or
                        i == 1 or
                        i == 2 or
                        i == 5 or
                        i == 14 or
                        i == 15 or
                        i == 16 or
                        i == 17
                    ):
                        head_elements_y_pos.append(point[1])

                    points.append(point)
                else:
                    points.append(None)

            # Draw Skeleton
            for num, pair in enumerate(POSE_PAIRS):
                partA = pair[0]
                partB = pair[1]
                if points[partA] and points[partB]:
                    cv2.line(
                        frame,
                        points[partA],
                        points[partB],
                        JOINT_COLORS[num],
                        3
                    )

            # Calculate head average position from its components
            if(len(head_elements_y_pos) > 0):
                head_avg_position = sum(head_elements_y_pos)
                head_avg_position /= len(head_elements_y_pos)
                # log.info(head_avg_position)

                # Compare previous head position
                # to detect if falling
                if (
                    previous_head_detection_frame and
                    (head_avg_position - previous_head_avg_position) >
                        fall_threshold and
                    (frame_number - previous_head_detection_frame) <
                        framerate_threshold
                ):
                    # print("Fall detected.")
                    last_fall_detected_frame = frame_number

                previous_head_avg_position = head_avg_position
                previous_head_detection_frame = frame_number

            # Draw Fall Detection Text if last fall event
            # ocurred max 2 seconds ago
            if (
                last_fall_detected_frame and
                (frame_number - last_fall_detected_frame) <= 2*fps
            ):
                cv2.putText(
                    frame,
                    "FALL DETECTED!",
                    fall_detected_text_position,
                    cv2.FONT_HERSHEY_COMPLEX, font_scale,
                    (0, 0, 255),
                    font_thickness,
                    cv2.LINE_AA
                )

        # If webcam mode
        if out_file:
            out_file.write(frame)
        else:
            cv2.imshow("Detection Results", frame)

        cur_request_id, next_request_id = next_request_id, cur_request_id
        frame = next_frame

        # Increment frame number
        frame_number += 1

        key = cv2.waitKey(1)
        if key == 27:
            break

    if out_file:
        # Release the out writer, capture, and destroy any OpenCV windows
        out_file.release()
        out_filename = os.path.splitext(input_stream)[0] + '_output.mp4'
        log.info("Finished. %s saved." % (out_filename))
    else:
        log.info("Finished.")
        cv2.destroyAllWindows()

    return ExitStatus.success

if __name__ == '__main__':
    sys.exit(main())
